### PR TITLE
Job Processor and Scaffolder Flow Implementation

### LINKED ISSUE #1108: Implement the Scaffolder Flow
## Feature Suggestion
- The scaffolder flow is currently implemented in a Mock example here https://github.com/spotify/backstage/blob/master/plugins/scaffolder-backend/src/service/router.ts#L40-L47

- This needs implementing properly.

## Possible Implementation
- A `scaffolder` is created which has 3 steps.
    - `Repository` - where to get the templates from.
    - `Templater` - the chosen templater (ie CookieCutter).
    - `VCS` - the interaction with the VCS (git/github).

As a simple implementation, when a `post` request is sent to the `router`, we could run all these steps and then send the response after all steps are complete, or return early if there are errors.

Moving forward, we can take a look at making these actual async jobs with the `201` response, and async workers picking up the templating jobs.
  
Depends on #1107 #1105 
Related to #345

### PR DESCRIPTION
**Pull request by benjdlambert**
- Scaffolder now can create jobs
- Will return an ID for the frontend to poll
- Supports grabbing all the logs from the Docker container that runs and any logging that happens in the Processor flow
- Still need to implement the VCS step but that's part of #1107 

#### :heavy_check_mark: Checklist

- [x] All tests are passing `yarn test`
- [ ] Screenshots attached (for UI changes)
- [ ] Relevant documentation updated
- [x] Prettier run on changed files
- [x] Tests added for new functionality
- [ ] Regression tests added for bug fixes

Fixes #1108

### CODE REVIEW

**Review by Rugvip — APPROVED** [review]
:itshappening:

File: plugins/scaffolder-backend/src/scaffolder/jobs/types.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+export type StageContext<T = {}> = {
+  values: RequiredTemplateValues & Record<string, JsonValue>;
+  entity: TemplateEntityV1alpha1;
+  logger: Logger;
+  logStream: Writable;
+} & T;
+
+export type ProcessorStatus = 'PENDING' | 'STARTED' | 'COMPLETED' | 'FAILED';
+
+export interface Stage extends StageInput {
```
[the remark this answers is not in the document]
Comment (benjdlambert) [reply plugins/scaffolder-backend/src/scaffolder/jobs/types.ts:34]: Will fix in next PR

File: plugins/scaffolder-backend/src/scaffolder/jobs/logger.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { PassThrough } from 'stream';
+import winston from 'winston';
+import { JsonValue } from '@backstage/config';
+
+export const useLogStream = (meta: Record<string, JsonValue>) => {
```
[the remark this answers is not in the document]
Comment (benjdlambert) [reply plugins/scaffolder-backend/src/scaffolder/jobs/logger.ts:20]: Will fix in next PR

File: packages/backend-common/src/logging/rootLogger.ts [hunk 3]
```diff
@@ -13,7 +13,6 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
```
Comment (freben) [inline packages/backend-common/src/logging/rootLogger.ts:16]: ಠ╭╮ಠ

File: packages/backend-common/src/middleware/errorHandler.ts [hunk 4]
```diff
@@ -53,7 +53,10 @@ export function errorHandler(
     next: NextFunction,
   ) => {
     if (response.headersSent) {
+      // If the headers have already been sent, do not send the response again
+      // as this will throw an error in the backend.
       next(error);
+      return;
```
Comment (freben) [inline packages/backend-common/src/middleware/errorHandler.ts:59]: ヽ(´▽`)/

File: plugins/scaffolder-backend/src/scaffolder/jobs/logger.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (freben) [reply plugins/scaffolder-backend/src/scaffolder/jobs/logger.ts:20]: also, should live in backend-common

**Pull request merged**